xpti/include/xpti/xpti_trace_framework.hpp

```diff
@@ -585,33 +597,162 @@ class tracepoint_t {
+      }
+    }
+  }
+  // Constructor that makes calls to xpti API layer to register strings and
+  // create the Universal ID that is stored in the TLS entry for lookup; this
+  // constructor is needed when only code location information is available
+  tracepoint_t(const char *fileName, const char *funcName, int line, int column,
+               void *codeptr = nullptr)
+      : m_payload(nullptr), m_top(false) {
+    // If tracing is not enabled, don't do anything
+    if (!xptiTraceEnabled())
+      return;
+    init();
+    printf("TLS Tracepoint@%s:%d\n", fileName ? fileName : "(Null)", line);
+
+    // Before we use the code location information, we need to check if a
+    // tracepoint has been set at a higher scope.
+    uint64_t uid = xptiGetUniversalId();
+    if (uid != xpti::invalid_uid) {
+      // We already have a parent SW layer that has a tracepoint defined. This
+      // should be associated with a trace event and a payload
+      m_trace_event =
+          const_cast<xpti::trace_event_data_t *>(xptiFindEvent(uid));
+      // If the trace event is valid, extract the payload
+      if (m_trace_event) {
+        m_payload = m_trace_event->reserved.payload;
       } else {
-        m_top = true;
-        uid = xptiRegisterPayload(p);
-        if (uid != xpti::invalid_uid) {
-          xptiSetUniversalId(uid);
-          m_payload = xptiQueryPayloadByUID(uid);
-        }
+        // Trace event is unavailable, so let is create one with the payload
+        // associated with the UID;
+        m_payload = xptiQueryPayloadByUID(uid);
+        m_trace_event = xptiMakeEvent(
+            m_default_name, const_cast<xpti::payload_t *>(m_payload),
+            m_default_event_type, m_default_activity_type, &m_instID);
+      }
+    } else if (fileName || funcName) {
+      // We expect the the file name and function name to be valid
+      m_top = true;
+      // Create a payload structure from the code location data
+      payload_t p(funcName, fileName, line, column, codeptr);
+      // Register the payload to generate the UID
+      uid = xptiRegisterPayload(&p);
+      if (uid != xpti::invalid_uid) {
+        xptiSetUniversalId(uid);
+        m_payload = xptiQueryPayloadByUID(uid);
+        m_trace_event = xptiMakeEvent(
+            m_default_name, const_cast<xpti::payload_t *>(m_payload),
+            m_default_event_type, m_default_activity_type, &m_instID);
       }
     }
   }
   ~tracepoint_t() {
+    // If tracing is not enabled, don't do anything
+    if (!xptiTraceEnabled())
+      return;
+
     if (m_top) {
       xptiSetUniversalId(xpti::invalid_uid);
     }
   }
 
+  tracepoint_t &stream(const char *stream_name) {
+    // If tracing is not enabled, don't do anything
+    if (xptiTraceEnabled()) {
+      m_default_stream = xptiRegisterStream(stream_name);
+    }
+    return *this;
+  }
+
+  tracepoint_t &trace_type(xpti::trace_point_type_t type) {
+    m_default_trace_type = (uint16_t)type;
```

Comment:
Why no `xptiTraceEnabled()` check?

Reply:
An if check may be more expensive than the simple statement?